### ISSUE TITLE
Investigate solr's hangups

### ISSUE THREAD

**Issue by jechols** [OregonDigital/OD2#826]
### Descriptive summary

@lsat12357 and @luisgreg99 have had trouble with the migration due to intermittent problems where solr just stops responding.  Nobody has been able to figure out what's going on.

### Expected behavior

Solr doesn't suddenly stop responding for no obvious reason.

**Comment by jechols** [comment]
Investigation notes so far:

- This is intermittent, but sometimes can be reproduced if the environment is backed up (I have a script that backs up the docker volumes and the local `./tmp` directory)
- @lsat12357 and @luisgreg99 have seen this ingesting collections as well as when migration collections' assets (I haven't gotten that far)
- I've had the most success in reproducing this by starting with a fresh stack, running the `firstrun.sh` script, and then running the rake task which generates our collections (`bundle exec rake oregon_digital:create_collections`).  Usually 15-30 collections in, solr stops responding.
- When this happens, Solr's CPU usage is zero, and its RAM is under 500 megs.  Swap is essentially unused, and I have plenty of idle CPU and unused RAM. (`docker stats` can help debug here)
- After killing the process which is talking to solr, it will eventually respond again, as if nothing happened
- While this is happening, Solr's web UI can be reached but won't do anything except complain that Solr is unavailable
- With logging turned way up (debug), I still don't see anything obvious that would explain what's going on.  Solr acts like nothing is wrong.
- Changing java's heap to 768mb, 1gb, and 2gb doesn't seem to change anything
- Solr version doesn't seem to matter.  This has been reproduced in solr 6 (alpine, slim, and normal) and solr 7 (slim and normal)
- `strace` on the java process didn't yield anything useful
  - I probably need to learn more about `strace` to get more from it, but I highly doubt that java or solr are the underlying problem (see above point about solr versions); we've used solr 6 for much more intense operations than loading 100-ish small objects and not seen anything remotely similar to this before

**Comment by jechols** [comment]
Bulk ingest problems' notes:

**tl;dr**: It may be that we have two separate problems happening, and the ingest problems could be easy to tackle just by having devs up the heap size.

One of the concerns with solr problems was during a bulk ingests.  I haven't been able to duplicate solr hanging during a bulk ingest, but ingesting PDFs tends to require a *lot* more RAM than solr typically needs.  It appears that we have an embedded Java app that deals with extracting text from various file types, including PDFs, Word documents, etc.

While 512m works fine for most stuff we do in dev (and for that matter for our Historic Oregon Newspapers production site), we need at least 2 gigs for ingesting documents that need text extraction.  It's almost tragic that we need so much extra RAM solely to handle ingesting a single type of asset, but it's possible there's nothing to be done about this without extracting text in a separate job, and while that would be a wonderful thing to pull into its own service, it's not likely worth the extra headaches of figuring that out and maintaining another weird brittle workflow.

**Comment by jechols** [comment]
Leaving this for next WC.  Solr problems were never fully figured out.  There's a branch we will want to use to see if it fixes problems, "feature/fix-solr", which radically changes our overall Solr configuration:

- Adjusts Solr's heap size for development and crashes on OOM so we can know exactly what happened and *when*
- Bumps dev and testing up to Solr 8.2
- Rewrites all Solr configuration more or less from scratch to merge the newest configs with the needs of Hyrax

So far testing is working well **for me**.  We'll need more testing next work cycle and then we can maybe deploy to staging.  And maybe... *just maybe*... this fixes some of our issues.

**Comment by jechols** [comment]
I think this can be closed - we've investigated THE HELL out of solr's hangups.  Whether or not we've got a solution is still yet to be determined, but for the moment I think there's not much more to say here.
